core/drivers/regulator/regulator_gpio.c

```diff
@@ -0,0 +1,261 @@
+					   bool enabled)
+{
+	struct regulator_gpio *regu = regulator_priv(regulator);
+
+	if (regu->enable_gpio) {
+		if (enabled) {
+			gpio_set_value(regu->enable_gpio, GPIO_LEVEL_HIGH);
+			if (regu->enable_delay)
+				udelay(regu->enable_delay);
+		} else {
+			gpio_set_value(regu->enable_gpio, GPIO_LEVEL_LOW);
+		}
+	}
+
+	return TEE_SUCCESS;
+}
+
+static TEE_Result regulator_gpio_read_state(struct regulator *regulator,
+					    bool *enabled)
+{
+	struct regulator_gpio *regu = regulator_priv(regulator);
+
+	if (regu->enable_gpio)
+		*enabled = gpio_get_value(regu->enable_gpio);
+	else
+		*enabled = true;
+
+	return TEE_SUCCESS;
+}
+
+static TEE_Result regulator_gpio_set_voltage(struct regulator *regulator,
+					     int level_uv)
+{
+	struct regulator_gpio *regu = regulator_priv(regulator);
+
+	if (level_uv == regu->voltage_levels_uv[0])
+		gpio_set_value(regu->voltage_gpio, GPIO_LEVEL_LOW);
+	else if (level_uv == regu->voltage_levels_uv[1])
+		gpio_set_value(regu->voltage_gpio, GPIO_LEVEL_HIGH);
+	else
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	return TEE_SUCCESS;
+}
+
+static TEE_Result regulator_gpio_read_voltage(struct regulator *regulator,
+					      int *level_uv)
+{
+	struct regulator_gpio *regu = regulator_priv(regulator);
+
+	*level_uv = regu->voltage_levels_uv[gpio_get_value(regu->voltage_gpio)];
+
+	return TEE_SUCCESS;
+}
+
+static const struct regulator_ops regulator_gpio_ops = {
+	.set_state = regulator_gpio_set_state,
+	.get_state = regulator_gpio_read_state,
+	.set_voltage = regulator_gpio_set_voltage,
+	.get_voltage = regulator_gpio_read_voltage,
+};
+
+static TEE_Result get_enable_gpio(const void *fdt, int node,
+				  struct regulator_gpio *regu)
+{
+	TEE_Result res = TEE_ERROR_GENERIC;
+	const fdt32_t *cuint = NULL;
+	struct gpio *gpio = NULL;
+
+	res = gpio_dt_get_by_index(fdt, node, 0, "enable", &gpio);
+	if (res == TEE_ERROR_ITEM_NOT_FOUND) {
+		regu->enable_gpio = NULL;
+
+		return TEE_SUCCESS;
+	}
+	if (res)
+		return res;
+
+	/* Override active level phandle flag, as per DT bindings */
+	if (dt_have_prop(fdt, node, "enable-active-high"))
+		gpio->dt_flags &= ~GPIO_ACTIVE_LOW;
+	else
+		gpio->dt_flags |= GPIO_ACTIVE_LOW;
+
+	/* Override open drain/open source phandle flag, as per DT bindings */
+	if (dt_have_prop(fdt, node, "gpio-open-drain"))
+		gpio->dt_flags |= GPIO_LINE_OPEN_DRAIN;
+	else
+		gpio->dt_flags &= ~GPIO_LINE_OPEN_DRAIN;
+
+	cuint = fdt_getprop(fdt, node, "startup-delay-us", NULL);
+	if (cuint)
+		regu->enable_delay = fdt32_to_cpu(*cuint);
+
+	gpio_set_direction(gpio, GPIO_DIR_OUT);
+
+	regu->enable_gpio = gpio;
+
+	return TEE_SUCCESS;
+}
+
+static TEE_Result get_voltage_level_gpio(const void *fdt, int node,
+					 struct regulator_gpio *regu)
+{
+	TEE_Result res = TEE_ERROR_GENERIC;
+	const fdt32_t *cuint = NULL;
+	struct gpio *gpio = NULL;
+	void *gpio_ref = &gpio;
+	int len = 0;
+
+	/*
+	 * DT bindings allows more than 1 GPIO to controller more than
```

Comment:
s/to controller/to control

maybe simplify to:
"DT bindings allows one GPIO to control ..."

Reply:
I fixed the typo but did not rephrase as suggested (I think it's less clear why the proposed implementation supports only 1 GPIO)

Reply:
Ok, no problem